CODE_REVIEW.md: how to do code reviews in curl (#5555)
curl/curl: docs/CODE_REVIEW.md

# How to do code reviews for curl

Anyone and everyone is encouraged and welcome to review code submissions in
curl. This is a guide on what to check for and how to perform a successful
code review.

## All submissions should get reviewed

> All pull requests and patches submitted ot the project should be reviewed by

Reviewer: s/ot/to/

at least one experienced curl maintainer before that code is accepted and

> received.

Reviewer: "received" sounds a bit weird, how about "merged into the main development branch"?

## Let the tools and tests take the first rounds

On initial pull requests, let the tools and tests do their job first and then

> start out by helping the submitted understand the test failures and tool

Reviewer: s/submitted/submitter/

alerts.

## How to provide feedback to author

> Be nice. Ask questions. Provide examples or suggestions of improvements.

Reviewer: "Assume best intentions."

All first-time contributors can become regulars. Let's help them go there.

## Is this a change we want?

If this is not a change that seems to be aligned with the project's path
forward and as such cannot be accepted, inform the author about this sooner
rather than later. Do it gently and explain why and possibly what could be
done to make it more acceptable.

## Code style

Most code style nits are detected by checksrc but not all. Only leave remarks

> on style deviation once checksrc doesn't find anymore.

Reviewer: "Minor nits from new submitters can also be handled by the maintainer when committing in case it seems like the submitter isn't clear on what to do, we want to make the process fun and exciting for new contributors."

## Encourage conformity

Make sure new code is written in a similar style as existing code. Naming,
logic, conditions, etc.

## Are pointers always non-NULL

If a function or code rely on pointers being non-NULL, take an extra look if
that's seems to be a fair assessment.

## Asserts

Conditions that should never be false can be verified with `DEBUGASSERT()`
calls to get caught in tests and debugging easier, while not having an impact
on final or release builds.

## Avoid and reduce mallocs

Can the mallocs be avoided? Do not introduce mallocs in any hot paths. If

> there are (new) mallocs, can they be combined into fewer calls?

Reviewer: "Are allocations handled in errorpaths to avoid/reduce leaks?"

## Should features be #ifdef'ed

Features and functionality may not be present everywhere and should therefore
be #ifdef'ed. Additionally, some features should be possible to switch on/off
in the build.

Write #ifdefs to be as little of a "maze" as possible.

## Does it look portable enough

curl runs "everywhere". Does the code take a reasonable stance and enough

> precautions to be possible to build and run on most platforms?

Reviewer: Should we mention sticking to C89 here? And perhaps endian-ness

## Tests and testability

New features should be added in conjunction with one or more test cases.
Ideally, functions should also be written so that unit tests can be done to
test invidual functions.

## Documentation

New features or changes to existing functionality **must** be accompanied with
updated documentation. Submitting that in a separate follow-up pull request is
not OK. A code review must also verify that the submitted documentation update

> matches the code submission.

Reviewer: I think we should add something about not having a too high bar for docs, but rather that it's a collaborative process: "English isn't everyone first language, be mindful of this and help the submitter improve the text if it needs a rewrite to read better."

# How to do code reviews for curl

Anyone and everyone is encouraged and welcome to review code submissions in
curl. This is a guide on what to check for and how to perform a successful

> code review.

Reviewer: I suggest that we add a blurb about how we really want anyone who is interested to join, not just C experts. I'm not sure how to word it, but perhaps something like "Code review is not exclusively for subject matter experts or curl maintainers, everyone at every level is welcome. If a PR isn't understandable then asking for clarification, or more code comments, is a perfectly valid review input."

Author: Isn't that just saying the same thing as *Anyone and everyone is encouraged and welcome to review code submissions in curl.* but with more words? Do we really need to spell it out like that?

Reviewer: Perhaps we don't no.

## All submissions should get reviewed

All pull requests and patches submitted to the project should be reviewed by
at least one experienced curl maintainer before that code is accepted and
merged.

## Let the tools and tests take the first rounds

On initial pull requests, let the tools and tests do their job first and then
start out by helping the submitter understand the test failures and tool
alerts.

## How to provide feedback to author

Be nice. Ask questions. Provide examples or suggestions of improvements.
Assume best intentions. Remember language barriers.

All first-time contributors can become regulars. Let's help them go there.

> ## Is this a change we want?

Reviewer: Maybe we should add a section on ABI changes somewhere around here? If the PR changes an exported prototype then it can easily go on the no-fly zone

Author: Yes, very good point. That's certainly something to be aware of and look for in a review.

Author: I'm adding this new section:
-----
## API/ABI stability or changed behavior

Changing the API and the ABI may be fine in a change but it needs to be done
deliberately and carefully. If not, a reviewer must help the author to realize
the mistake.

curl and libcurl are similarly very strict on not modifying existing
behavior. API and ABI stability is not enough, the behavior should also remain
intact as far as possible.

Reviewer: LGTM

Reviewer 2: Maybe this could contain some more info about what API/ABI compatibility means?

For example: I plan to move some select/timeval related code to curlx functions in the future, but I am not yet sure how to do that, because I don't know how new functions affect API/ABI compatibility. On Windows it is probably quite easy, because (DLL) functions need to explicitly exported, but on Linux this is not the case, right?

Author: The ABI guideline is [here](https://github.com/curl/curl/blob/b71628b633de3287fa83360a2019f0b7023b3aa3/docs/libcurl/ABI) (and I just realized I should convert that to markdown) - I think it's better to keep that separate, but of course we also need to make sure it is accurate and helpful enough.

If this is not a change that seems to be aligned with the project's path
forward and as such cannot be accepted, inform the author about this sooner
rather than later. Do it gently and explain why and possibly what could be
done to make it more acceptable.

## Code style

Most code style nits are detected by checksrc but not all. Only leave remarks

> on style deviation once checksrc doesn't find anymore.

Reviewer 3: ```suggestion
on style deviation once checksrc doesn't find any more.
```

Author: I'm not sure I agree with this (but English is not my native language so I'm willing to get corrected on this). I intended the meaning as in "when it can't find problems any longer" which is typically how "anymore" differs from "any more" - as the latter is more about any further instances. One source on the difference: https://www.grammarly.com/blog/anymore-vs-any-more/

Reviewer 4: It should be "any more" as in "any more errors"

Author: Thank you.

Minor nits from fresh submitters can also be handled by the maintainer when
merging, in case it seems like the submitter isn't clear on what to do. We
want to make the process fun and exciting for new contributors.

> ## Encourage conformity

Reviewer 4: Perhaps consistency is a better word than conformity.

Make sure new code is written in a similar style as existing code. Naming,
logic, conditions, etc.

> ## Are pointers always non-NULL

Reviewer 3: ```suggestion
## Are pointers always non-NULL?
```

If a function or code rely on pointers being non-NULL, take an extra look if

> that's seems to be a fair assessment.

Reviewer 3: ```suggestion
that seems to be a fair assessment.
```

## Asserts

Conditions that should never be false can be verified with `DEBUGASSERT()`
calls to get caught in tests and debugging easier, while not having an impact
on final or release builds.

## Memory allocation

Can the mallocs be avoided? Do not introduce mallocs in any hot paths. If
there are (new) mallocs, can they be combined into fewer calls?

Are all allocations handled in errorpaths to avoid leaks and crashes?

> ## Should features be #ifdef'ed

Reviewer 3: ```suggestion
## Should features be `#ifdef`ed?
```

Features and functionality may not be present everywhere and should therefore

> be #ifdef'ed. Additionally, some features should be possible to switch on/off

Reviewer 3: ```suggestion
be `#ifdef`ed. Additionally, some features should be possible to switch on/off
```

in the build.

> Write #ifdefs to be as little of a "maze" as possible.

Reviewer 3: ```suggestion
Write `#ifdef`s to be as little of a "maze" as possible.
```

> ## Does it look portable enough

Reviewer 3: ```suggestion
## Does it look portable enough?
```

curl runs "everywhere". Does the code take a reasonable stance and enough
precautions to be possible to build and run on most platforms?

Remember that we live by C89 restrictions.

## Tests and testability

New features should be added in conjunction with one or more test cases.
Ideally, functions should also be written so that unit tests can be done to
test individual functions.

## Documentation

New features or changes to existing functionality **must** be accompanied with
updated documentation. Submitting that in a separate follow-up pull request is
not OK. A code review must also verify that the submitted documentation update
matches the code submission.

> English isn't everyone first language, be mindful of this and help the

Reviewer 3: ```suggestion
English isn't everyone's first language, be mindful of this and help the
```

Source code should be written to maximize readability and be easy to
understand.

## Functions shouldn't be large

A single function should never be large as that makes it hard to follow and
understand all the exit points and state changes. Some existing functions in
curl certainly violate this ground rule but when reviewing new code we should
propose splitting into smaller functions.

## Duplication is evil

Anything that looks like duplicated code is a red flag. Anything that seems to
introduce code that we *should* already have or provide needs a closer check.

## Sensitive data

When credentials are involved, take an extra look at what happens with this
data. Where it comes from and where it goes.

## Variable types differ

`size_t` is not a fixed size. `time_t` can be signed or unsigned and have
different sizes. Relying on variable sizes is a red flag.

Also remember that endianness and >= 32 bit accesses to unaligned addresses
are problematic areas.

## Integer overflows

Be careful about integer overflows. Some variable types can be either 32 bit
or 64 bit. Integer overflows must be detected and acted on *before* they
happen.

## Dangerous use of functions

> Maybe use of realloc() should rather use the dynbuf functions?

Reviewer 3: ```suggestion
Maybe use of `realloc()` should rather use the dynbuf functions?
```

# How to do code reviews for curl

Anyone and everyone is encouraged and welcome to review code submissions in
curl. This is a guide on what to check for and how to perform a successful
code review.

## All submissions should get reviewed

All pull requests and patches submitted to the project should be reviewed by
at least one experienced curl maintainer before that code is accepted and

> merged.

Reviewer 2: Is this also the case for core team members? I am currently thinking about this like:

If the change is simple, in line with existing things and has been tested by the core team member, the member can either directly push to master or create a PR for CI testing purposes, but it does not require a review.

Author: I went with "should" here as I believe this is what we *should* do. Persons with "merge access" in the project can always make the judgement call themselves if that is done or not, and I also didn't say that you can't review your own code. I believe **I** am one of the most prolific submit-PR-then-merge persons who land PRs without any second person having chimed in on it - and I believe I do that for the good of the project, not because I fear other people's comments - basically moving on when I think it is good enough after having given others a chance to comment/object/praise.

I still claim that *"at least one experienced curl maintainer should do a review before that code is accepted"*

Reviewer 2: Okay, I am with you on this. I just did not think reviewing the PR yourself and giving others maybe a few hours to chime in fulfills this sentence, so maybe this could be made more explicit?

Author: I don't want to document shortcuts as it would sound like we encourage or suggest that to be a way we think is good.

I prefer to document how we *should* do and act. And then individuals with the given "authority" can decide to side-step the recommendations if they think it is beneficial for that particular case.

Reviewer 2: > it would sound like we encourage or suggest that to be a way we think is good.

But doesn't this imply that not having at least one review is bad practice and discouraged?

Author: Yes!

Reviewer 2: > And then individuals with the given "authority" can decide to side-step the recommendations if they think it is beneficial for that particular case.

Maybe we need a separate document describing what members are allowed to do. I would like to play by the rules which I am not sure I am up-to-date with.

Author: I think the [GOVERNACE](https://github.com/curl/curl/blob/master/docs/GOVERNANCE.md) document should be the one that talks about who is allowed to do what in the project in general.

Imagine a cookbook. It shows the recipe exactly how it should be done so that people know exactly how to do it, in particular if they never did these cookies before. But once you've baked those cookies many times, you might be allowed to take some liberties and freestyle a bit - for the sake of effectiveness and performance. That's wiggle room.

The cookbook doesn't document how or when to make shortcuts - and it rightfully doesn't even suggest that they are a good idea. That's something for every cook to do, or not to, according to your own judgement, skill and comfort-level. Going off-recipe is of course not recommended unless you really think you can still deliver at least the same quality end product. 

"freestyling" is done on your own responsibility and with the full awareness that it isn't by the book anymore and therefor I can rightfully be blamed if I burn down the house in the process.

I don't want to document that we're not allowed to burn down the house.

Reviewer 2: Thanks for the detailed explanation!

## Let the tools and tests take the first rounds

On initial pull requests, let the tools and tests do their job first and then
start out by helping the submitter understand the test failures and tool
alerts.

## How to provide feedback to author

Be nice. Ask questions. Provide examples or suggestions of improvements.
Assume best intentions. Remember language barriers.

All first-time contributors can become regulars. Let's help them go there.

## Is this a change we want?

> If this is not a change that seems to be aligned with the project's path

Reviewer 5: unsure if we need a generic 'preflight' checklist when doing a code review ... of which we could highlight/require that every PR needs to reference an issue.

Reviewer 2: What about tidy-up PRs? Would these require an issue?

Reviewer 5: I do not think we need to say ***must*** eg. PR's that do not reference an issue # should be rare.

Author: If we should say something about issues, it should then probably be something about the commit message - which the current version of the document skips completely. I don't think we will benefit from requiring an issue to exist before a pull-request is made. Looking back at how we typically work, a rather large portion of our bugfixes are done as a PR right away, shortcutting over the issue if the person who does the fix also is the person who found the issue.

Reviewer 5: code reviewer should encourage contributors to scan existing issue list for previous issues and try to link PR/commit with an issue ... not suggesting we need to backfill issues to satisfy.

Author: Agreed, but technically speaking the commit message is not code so it's not part of a "code review". It is however part of the "PR review" or perhaps the things-to-check-before-merge. What do you say about an added section like below?

---

# Commit messages

Tightly coupled with a code review is making sure that the commit message is
good. It is the responsibilitiy of the person who merges the code to make sure
that the commit message follows our standard (detailed in the
[CONTRIBUTE.md](CONTRIBUTE.md) document). This includes making sure the PR
identifies related issues and giving credit to reporters and helpers.

forward and as such cannot be accepted, inform the author about this sooner
rather than later. Do it gently and explain why and possibly what could be
done to make it more acceptable.

## API/ABI stability or changed behavior

Changing the API and the ABI may be fine in a change but it needs to be done
deliberately and carefully. If not, a reviewer must help the author to realize
the mistake.

curl and libcurl are similarly very strict on not modifying existing
behavior. API and ABI stability is not enough, the behavior should also remain
intact as far as possible.

## Code style

Most code style nits are detected by checksrc but not all. Only leave remarks
on style deviation once checksrc doesn't find any more.

Minor nits from fresh submitters can also be handled by the maintainer when
merging, in case it seems like the submitter isn't clear on what to do. We
want to make the process fun and exciting for new contributors.

## Encourage consistency

> Make sure new code is written in a similar style as existing code. Naming,

Reviewer 5: maybe reinforce how important it is to pick good names ....

Author: Feels like a tricky value judgement that is hard to write here, doesn't it? What's a "good name" other than being consistent with what we've done before? (with the hope that we've done decently in the past)

Reviewer 5: I am not suggesting to attempt explicitly constrain/define what a good name is ... I am saying making sure a PR has good names is what a code reviewer should be doing.

Author: Isn't that what is already stated? Do we need to spell it out more?

introduce code that we *should* already have or provide needs a closer check.

## Sensitive data

When credentials are involved, take an extra look at what happens with this
data. Where it comes from and where it goes.

## Variable types differ

`size_t` is not a fixed size. `time_t` can be signed or unsigned and have
different sizes. Relying on variable sizes is a red flag.

Also remember that endianness and >= 32 bit accesses to unaligned addresses
are problematic areas.

## Integer overflows

Be careful about integer overflows. Some variable types can be either 32 bit
or 64 bit. Integer overflows must be detected and acted on *before* they
happen.

## Dangerous use of functions

Maybe use of `realloc()` should rather use the dynbuf functions?

Do not allow new code that grows buffers without using dynbuf.

Use of C functions that rely on a terminating zero must only be used on data
that really do have a zero terminating zero.

## Dangerous "data styles"

Make extra precautions and verify that memory buffers that need a terminating
zero always have exactly that. Buffers *without* a zero terminator must not be

> used as input to string functions.

Reviewer 5: maybe add a section on multithreading/parallel gotchas

Reviewer 5: maybe we need a section on scope/arch as well eg. PR's should try to identify scope and be explicit in what work is being done vs work undone ... often people working in their small corner are focused on whats important to them (which is entirely fine) but in the course of their work rediscover larger refactor. 

AS for matters of arch ... @bagder creates a new spec like document which defines general approach ...that seems like something lightweight enough to recommend trying.

Author: Any example of phrasing in mind? There's not a lot of libcurl code for what you typically need to consider multi-threading, by design.

Reviewer 6: static variables

Author: Adding:

----
## Thread-safety

We do not like static variables as they break thread-safety and prevent functions from being reentrant.

All pull requests and patches submitted to the project should be reviewed by
at least one experienced curl maintainer before that code is accepted and
merged.

## Let the tools and tests take the first rounds

On initial pull requests, let the tools and tests do their job first and then
start out by helping the submitter understand the test failures and tool
alerts.

## How to provide feedback to author

Be nice. Ask questions. Provide examples or suggestions of improvements.
Assume best intentions. Remember language barriers.

All first-time contributors can become regulars. Let's help them go there.

## Is this a change we want?

If this is not a change that seems to be aligned with the project's path
forward and as such cannot be accepted, inform the author about this sooner
rather than later. Do it gently and explain why and possibly what could be
done to make it more acceptable.

## API/ABI stability or changed behavior

Changing the API and the ABI may be fine in a change but it needs to be done
deliberately and carefully. If not, a reviewer must help the author to realize
the mistake.

curl and libcurl are similarly very strict on not modifying existing

> behavior. API and ABI stability is not enough, the behavior should also remain

Reviewer 2: Maybe link to the API / ABI document from here then?

Author: I don't want to link across directories using markdown as that won't work when this is rendered on the curl web site as compared to rendered on github. On the web site we can have "related" link in the top right box that makes sense.
